src/MongoDB.Driver/Linq/FieldExpressionFlattener.cs

```diff
@@ -86,5 +80,43 @@ protected internal override Expression VisitField(FieldExpression node)
 
             return node;
         }
+
+        // private methods
+        private static string FormatArrayIndex(object index)
+        {
+            if (index is int intIndex)
+            {
+                return FormatArrayIndex((long)intIndex);
+            }
+
+            if (index is long longIndex)
+            {
+                return FormatArrayIndex(longIndex);
+            }
+
+            throw new NotSupportedException("Array indexes must be int or long.");
+        }
+
+        private static string FormatArrayIndex(long index) 
+        {
+            switch (index)
+            {
+                // We've treated -1 as meaning $ operator. We can't break this now,
+                // so, specifically when we are flattening fields names, this is 
+                // how we'll continue to treat -1.
+                case var _ when index < -1L:
+                {
+                    throw new NotSupportedException("Array indexes must be more or equal than -1.");
```

Comment:
This message is not 100% grammatically correct. I had suggested in a comment below:

    throw new NotSupportedException("Array indexes cannot be negative.");

I'm not actually 100% sure that `NotSupportedException` is the right exception. Can you think of any more suitable exception? If not, then good.

Reply:
It's difficult to say since the indexer allows (in other words - `supports`) a special index (-1), so all other negative special cases can be considered as `unsupported`. But, the same time 
`IndexOutOfRangeException` can be used here too.  
I've changed it.

I don't think that we can use this message: `Array indexes cannot be negative.` since we allow to have '-1'. Maybe something like this: `"Array indexes must be greater or equal than -1."` ?

Reply:
If ignored -1 on purpose, but if you want a more precise message the grammatically correct form would be:

    Array indexes must be greater than or equal to -1.

Reply:
Done